core/src/main/java/io/grpc/ChannelImpl.java

```diff
@@ -332,7 +329,7 @@ public void sendPayload(ReqT payload) {
       // For unary requests, we don't flush since we know that halfClose should be coming soon. This
       // allows us to piggy-back the END_STREAM=true on the last payload frame without opening the
       // possibility of broken applications forgetting to call halfClose without noticing.
-      if (!unaryRequest) {
+      if (!method.getType().clientSendsOneMessage()) {
```

Comment:
rather than check that client doesn't send one message, I'd suggest the following to avoid the negated ```if``` check

``` if (method.getType().clientSendsMultipleMessages()) { ```